java/fury-core/src/main/java/org/apache/fury/io/FuryReadableChannel.java

```diff
@@ -20,30 +20,130 @@
+    this.memoryBuffer = new MemoryBuffer(offHeapAddress, 0, directBuffer, this);
+  }
+
+  @Override
+  public int fillBuffer(int minFillSize) {
+    try {
+      ByteBuffer byteBuf = byteBuffer;
+      MemoryBuffer memoryBuf = memoryBuffer;
+      int position = byteBuf.position();
+      int newLimit = position + minFillSize;
+      if (newLimit > byteBuf.capacity()) {
+        int newSize =
+            newLimit < BUFFER_GROW_STEP_THRESHOLD ? newLimit << 2 : (int) (newLimit * 1.5);
+        ByteBuffer newByteBuf = ByteBuffer.allocateDirect(newSize);
+        byteBuf.position(0);
+        newByteBuf.put(byteBuf);
+        byteBuf = byteBuffer = newByteBuf;
+        memoryBuf.initDirectBuffer(Platform.getAddress(byteBuf), position, byteBuf);
+      }
+      byteBuf.limit(newLimit);
+      int readCount = channel.read(byteBuf);
+      memoryBuf.increaseSize(readCount);
+      return readCount;
+    } catch (IOException e) {
+      throw new DeserializationException("Failed to read the provided byte channel", e);
+    }
   }
 
   @Override
   public int read(ByteBuffer dst) throws IOException {
-    throw new UnsupportedEncodingException();
+    int dstRemaining = dst.remaining();
+    if (dstRemaining <= 0) {
+      return 0;
+    }
+    MemoryBuffer buf = memoryBuffer;
+    int remaining = buf.remaining();
+    if (remaining <= 0) {
+      return -1;
+    }
+    if (remaining >= dstRemaining) {
+      byte[] bytes = buf.readBytes(dstRemaining);
+      dst.put(bytes);
+      return dstRemaining;
+    } else {
+      int filledSize = fillBuffer(dstRemaining - remaining);
+      int length = remaining + filledSize;
+      byte[] bytes = buf.readBytes(length);
+      dst.put(bytes);
+      return length;
+    }
+  }
+
+  @Override
+  public void readTo(byte[] dst, int dstIndex, int length) {
+    MemoryBuffer buf = memoryBuffer;
+    int remaining = buf.remaining();
+    if (remaining >= length) {
+      buf.readBytes(dst, dstIndex, length);
+    } else {
+      buf.readBytes(dst, dstIndex, remaining);
+      try {
+        ByteBuffer buffer = ByteBuffer.wrap(dst, dstIndex + remaining, length - remaining);
+        channel.read(buffer);
+      } catch (IOException e) {
+        throw new DeserializationException("Failed to read the provided byte channel", e);
+      }
+    }
+  }
+
+  @Override
+  public void readToUnsafe(Object target, long targetPointer, int numBytes) {
+    MemoryBuffer buf = memoryBuffer;
+    int remaining = buf.remaining();
+    if (remaining < numBytes) {
+      fillBuffer(numBytes - remaining);
+    }
+    long address = buf.getUnsafeReaderAddress();
+    Platform.copyMemory(null, address, target, targetPointer, numBytes);
+    buf.increaseReaderIndex(numBytes);
+  }
+
+  @Override
+  public void readToByteBuffer(ByteBuffer dst, int length) {
+    MemoryBuffer buf = memoryBuffer;
+    int remaining = buf.remaining();
+    if (remaining < length) {
+      remaining += fillBuffer(length - remaining);
```

Comment:
This introduce an extra copy, the data doesn't have to copy into `memoryBuffer` to read into `dst`

Reply:
I thought we need it, to have the ability to *go back*.
If we don't, to avoid using some transfer buffer we'll be to set a limit to the destination buffer. Is this acceptable?

Reply:
Actually I'm considering only provide streaming interface to customized serializer to forbidden users to use API that can go back